implement `preimage` for date_trunc

drin opened the pull request:
Originally, this attempted to implement a custom optimizer rule in the datafusion expression simplifier. Now, this has been updated to work within the new preimage framework rather than being implemented directly in the expression simplifier.

## Which issue does this PR close?
Closes #18319.

## Rationale for this change
To transform binary expressions that compare `date_trunc` with a constant value into a form that can be better utilized (improved performance).

For Bauplan, we can see the following (approximate average over a handful of runs):

Q1:
```sql
SELECT PULocationID, trip_miles, tips
  FROM taxi_fhvhv
 WHERE date_trunc('month', pickup_datetime) <= '2025-01-08'::DATE
```

Q2:
```sql
SELECT PULocationID, trip_miles, tips
  FROM taxi_fhvhv
 WHERE pickup_datetime < date_trunc('month', '2025-02-08'::DATE)
```

|  Query  | Time (s) | Options |
| ------- | -------- | -------- |
|    Q1     |    ~3    |  no cache, optimization enabled |
|    Q1     |   ~35   |  no cache, optimization disabled |
|    Q2     |    ~3    |  no cache, optimization enabled |
|    Q2     |    ~3    |  no cache, optimization disabled |

## What changes are included in this PR?
A few additional support functions and additional match arms in the simplifier match expression.

## Are these changes tested?

Our custom rule has tests of the expression transformations and for correct evaluation results. These will be added to the PR after the implementation is in approximately good shape.

## Are there any user-facing changes?
Better performance and occasionally confusing explain plan. In short, a `date_trunc('month', col) = '2025-12-03'::DATE` will always be false (because the truncation result can never be a non-truncated value), which may produce an unexpected expression (`false`).

Explain plan details below (may be overkill but it was fun to figure out):

Initial query:
```sql
SELECT  PULocationID
           ,pickup_datetime
      FROM taxi_view_2025
     WHERE date_trunc('month', pickup_datetime) = '2025-12-03'
```

After simplify_expressions:
```sql
logical_plan after simplify_expressions                    | Projection: taxi_view_2025.PULocationID, taxi_view_2025.pickup_datetime                                                                                            |
|                                                            |   Filter: date_trunc(Utf8("month"), CAST(taxi_view_2025.pickup_datetime AS Timestamp(Nanosecond, None))) = TimestampNanosecond(1764720000000000000, None)          |
|                                                            |     TableScan: taxi_view_2025
```

Before and after `date_trunc_optimizer` (our custom rule):
```sql
logical_plan after optimize_projections                    | Filter: date_trunc(Utf8("month"), CAST(taxi_view_2025.pickup_datetime AS Timestamp(Nanosecond, None))) = TimestampNanosecond(1764720000000000000, None)            |
|                                                            |   TableScan: taxi_view_2025 projection=[PULocationID, pickup_datetime]                                                                                             |
| logical_plan after date_trunc_optimizer                    | Filter: Boolean(false)                                                                                                                                             |
|                                                            |   TableScan: taxi_view_2025 projection=[PULocationID, pickup_datetime]
```

drin committed:
feat: Implemented preimage for date_trunc

This implementation leverages `general_date_trunc` to truncate the
preimage input value and then uses the input granularity to increment
the truncated datetime by 1 unit.

drin committed:
test: added sql logic test for date_trunc preimage

This adds sql logic test for date_trunc preimage for test coverage

drin committed:
fix: changed function comments

This is to fix `DateTime<Tz>` being considered to include HTML rather
than a templated type. Also improved the phrasing of the comment

sdf-jkl commented on datafusion/functions/src/datetime/date_trunc.rs:
```diff
@@ -412,6 +415,151 @@ impl ScalarUDFImpl for DateTruncFunc {
+            .as_literal()
+            .and_then(|sv| sv.try_as_str().flatten())
+            .map(part_normalization);
+
```
We should add a guard for Type families. `col_expr: TimeStamp` needs a `lit_expr: TimeStamp` and the same for `Time` types.

sdf-jkl reviewed (commented):
@drin thanks for the wait.

I left some suggestions.

drin commented on datafusion/sqllogictest/test_files/datetime/timestamps.slt:
```diff
@@ -5358,3 +5358,301 @@ SELECT to_timestamp(arrow_cast(100.5, 'Float16'), name) FROM test_to_timestamp_s
+# Test YEAR granularity - basic comparisons
+
+query P
+SELECT ts FROM t1 WHERE date_trunc('year', ts) = timestamp '2024-01-01T00:00:00' ORDER BY ts;
```
(but, from those notes I also realized I was handling the interval wrong in some cases)

drin replied:
But these are the notes relevant to this specific case (predicate operator is `=`):
```
// Special condition:
// if date_trunc(part, const_rhs) != const_rhs, this is always false
// For this operator, truncation means that we check if the column is INSIDE of a range.
// lhs(=) --> column >= date_trunc(part, const_rhs) AND column < next_interval(part, const_rhs)
```

But, without knowing what the operator is, I can't know that I should return `None` from `preimage`. I can only return the interval. But, in `rewrite_with_preimage`, you can do the appropriate check:
```
Operator::Eq && lower != <original> => False,
Operator::Eq => and(<check if within interval>),
```

I'm not sure if this makes sense for intervals from non-truncating functions. I'd have to simmer on that

sdf-jkl replied:
I think it only makes sense for truncating functions. Right now, `rewrite_with_preimage` is function agnostic and is creation the expression based only on the `Operator` and `Interval`.

drin committed:
feat: expanded rewrite_with_preimage

This expands the predicate operator handling for rewrite_with_preimage
to accommodate boundary cases

drin committed:
fix: update date_part and floor

These updates are to accommodate the new `is_boundary` condition in
`PreimageResult`

drin committed:
feat: bug fixes and cleanup

This accommodates some review feedback and also fixes some edge cases
involving boundary conditions for preimage

drin commented on datafusion/functions/src/datetime/date_trunc.rs:
```diff
@@ -502,6 +650,125 @@ fn truncate_time_secs(value: i32, granularity: DateTruncGranularity) -> i32 {
+        DateTruncGranularity::Hour => value + MILLIS_PER_HOUR,
+        DateTruncGranularity::Minute => value + MILLIS_PER_MINUTE,
+        DateTruncGranularity::Second => value + MILLIS_PER_SECOND,
+        DateTruncGranularity::Millisecond => value + 1,
```
I don't think interval semantics should be handled here though (I'll fix it at the call site)

drin committed:
feat: accommodate PR feedback

This accommodates feedback about:
- Abstracting a closure from within `preimage`
- Reusing `valid_with_time`
- Possibly producing empty intervals (lower == upper)

This also adds a unit test to accommodate some cases that are actually
impossible except from direct invocation.

drin committed:
chore: ran cargo fmt

sdf-jkl commented:
I guess the first question is: do we really want to support edge cases like `date_trunc('month', '2026-02-27') < '2026-02-23'`?

Why not fall back to not using preimage, like the `floor` function? What arguments in favor of changing the API?

One argument I can think of is:
- Right now, it's up to the user to know not to use dates like `2026-02-XX` (where `XX > 01`). Leaving that responsibility to the user is not good UX. I'd be confused if some queries where the RHS day is `01` were faster than otherwise identical queries with a different day value.

On the other hand:
- API change and extra complexity (not that big of a change and doesn't add much complexity to the interval calculation)

sdf-jkl commented:
I'm convinced. We could split the PR into two smaller ones to make review easier:
1) Apply changes to the `preimage` api and the functions that already implement it.
2) Implement `preimage` for `date_trunc`

Let me know if you think that's a good idea.

sdf-jkl commented:
From my experience waiting for PR reviews, it’s often easier to review several smaller PRs than one large one. 

I ended up splitting my [original](https://github.com/apache/datafusion/pull/18789) `date_part` `preimage` PR into two. 

You could keep this one as is and add a new small one with the `preimage` API update.

alamb commented on datafusion/expr/src/preimage.rs:
```diff
@@ -23,7 +23,23 @@ use crate::Expr;
     /// No preimage exists for the specified value
     None,
-    /// The expression always evaluates to the specified constant
-    /// given that `expr` is within the interval
```
I don't fully understand this logic 

For the example `floor(x) < 8.3`  I would expect there to be no preimage as defined here -- specifically there is no range of inputs for which `floor(x)` evaluates to `8.3`

I agree that it is valid simplificaition to rewrite `floor(x) < 8.3` to `x < 9.0`, but it seems different than "preimage" 🤔 

Maybe we just need to give it a different name (maybe that is what you have tried to do with `is_boundary`)

alamb reviewed (commented):
After reading this PR and the conversations on it, I think it might help to break this down into smaller parts. 

1. Cases where we can clearly apply the (existing) preimage rewrite -- e.g. `date_trunc(date_col, 'month') = '2025-01-01'` where there is a range of `date_col` that always evaluates to `2025-01-01`
2. Other potential simplifications that don't clearly fall into the "preimage" category (e.g. `date_trunc(date_col, 'month') = '2025-01-02'` which will never be true)

I actually think the first one will be fairly straightforward and a clear win, though it will miss some potential predicates like `date_trunc(date_col, 'month') < '2025-01-02'`

As you have explored in this PR, the existing API for preimage is not sufficient for the second type of rewrite, and I think adding a new API would be eaiser to reason about in a follow on PR
